pkg/reaper/pdbreaper/pdbreaper.go

```diff
@@ -0,0 +1,383 @@
+
+		// check if multiple PDBs in a namespace contain refernece to same pods
+		for _, pdb := range pdbs {
+			log.Infof("evaluating multi-namespace PDB %v", pdbNamespacedName(pdb))
+
+			labelSelector, err := common.GetSelectorString(pdb.Spec.Selector)
+			if err != nil {
+				return errors.Wrapf(err, "failed to get label selector from structured selector %+v", pdb.Spec.Selector)
+			}
+
+			pods, err := ctx.listPodsWithSelector(namespace, labelSelector)
+			if err != nil {
+				return errors.Wrap(err, "failed to list PDB pods")
+			}
+
+			namespacePodsWithBudget = append(namespacePodsWithBudget, pods...)
+		}
+
+		if isContainDuplicatePods(namespacePodsWithBudget) {
+			log.Infof("PDBs %+v are marked reapable - pods %+v has multiple PDBs", pdbSliceNamespacedNames(pdbs), podSliceNamespacedNames(namespacePodsWithBudget))
+			ctx.addReapablePodDisruptionBudget(pdbs...)
+			for _, pdb := range pdbs {
+				err := ctx.publishEvent(pdb, EventReasonMultipleDetected, EventMessageMultipleFmt)
+				if err != nil {
+					log.Warnf(err.Error())
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (ctx *ReaperContext) listPodsWithSelector(namespace, selector string) ([]corev1.Pod, error) {
+	var pods []corev1.Pod
+	podList, err := ctx.KubernetesClient.CoreV1().Pods(namespace).List(metav1.ListOptions{LabelSelector: selector})
+	if err != nil {
+		return pods, errors.Wrapf(err, "failed to list pods with selector '%v'", selector)
+	}
+	pods = append(pods, podList.Items...)
+	return pods, nil
+}
+
+func (ctx *ReaperContext) publishEvent(pdb policyv1beta1.PodDisruptionBudget, reason, msg string) error {
+	var (
+		pdbNamespace   = pdb.GetNamespace()
+		pdbName        = pdb.GetName()
+		namespacedName = pdbNamespacedName(pdb)
+	)
+
+	now := time.Now()
+	event := &corev1.Event{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("pdb-reaper-%v", pdbName),
+			Namespace:    pdbNamespace,
+		},
+		InvolvedObject: corev1.ObjectReference{
+			Kind:            "PodDisruptionBudget",
+			Namespace:       pdbNamespace,
+			Name:            pdbName,
+			APIVersion:      pdb.APIVersion,
+			UID:             pdb.UID,
+			ResourceVersion: pdb.ResourceVersion,
+		},
+		Reason:         reason,
+		Message:        fmt.Sprintf(msg, namespacedName),
+		Type:           "Normal",
+		FirstTimestamp: metav1.NewTime(now),
+		LastTimestamp:  metav1.NewTime(now),
+	}
+	_, err := ctx.KubernetesClient.CoreV1().Events(pdbNamespace).Create(event)
+	if err != nil {
+		return errors.Wrap(err, "failed to publish event")
+	}
+	return nil
+}
+
+func (ctx *ReaperContext) addReapablePodDisruptionBudget(pdb ...policyv1beta1.PodDisruptionBudget) {
+	for _, p := range ctx.ReapablePodDisruptionBudgets {
+		if reflect.DeepEqual(p, pdb) {
+			return
+		}
+	}
+
+	ctx.ReapablePodDisruptionBudgetsCount += len(pdb)
+	ctx.ReapablePodDisruptionBudgets = append(ctx.ReapablePodDisruptionBudgets, pdb...)
+}
+
+func isMisconfigured(pdb policyv1beta1.PodDisruptionBudget, pods []corev1.Pod) (bool, error) {
```

Comment:
We might reconsider the criteria for a misconfigured PDB. A Job, BatchJob or short-lived Pod might configure a PDB for itself to make sure it is not interrupted and allowed to run to completion. We might want to add a qualifier to label "misconfigured" PDBs and only delete if they are considered "misconfigured" over several runs of the pdbreaper. IOWs, it's fine if the PDB blocks drains for some configurable number of minutes, but after that threshold it's no longer considered a short-lived Pod and the PDB should be reaped.

Reply:
If you know the PDB is misconfigured e.g. `maxUnavailable: 0`, why wait? PDBs are immutable anyway.
For the possibly temporary case of AllowedDisruptions: 0, we don't do anything if the PDB is configured properly.

Reply:
Setting `maxUnavailable: 0` might be acceptable for the use-case where a Pod is critical but relatively short-lived. Like an Argo workflow step or Jenkins build job. It might be better for the system as a whole to make node drains wait for those short-lived Pods to complete. It's when the PDB for a long-running web service has `maxUnavailable: 0` that there is a problem.

Reply:
Yes, in that case having time would be ideal, but unfortunately PDBs do not reference the time they became 'blocking', so like you mentioned the only way would be to do something like annotate the first time we noticed it and wait for N iterations before we delete.